reconciler/testing/reactions.go

```diff
@@ -48,6 +48,12 @@ func InduceFailure(verb, resource string) clientgotesting.ReactionFunc {
 
 func ValidateCreates(ctx context.Context, action clientgotesting.Action) (handled bool, ret runtime.Object, err error) {
 	got := action.(clientgotesting.CreateAction).GetObject()
+	o, ok := got.(apis.Defaultable)
+	if ok {
+		o.SetDefaults(ctx)
+		got, _ = o.(runtime.Object)
+	}
+
```

Comment:
Another alternative is to have a Defaulting reactor that calls SetDefaults on defaultable objects that are created as well as updated and in eventing we can choose to opt-in the SetDefault call by using it only for the sugar reconciler.

Reply:
I explored the Reactor route as well. But in eventing, we use [prependReactors](https://github.com/knative/eventing/blob/abe0a570af624f98b9c2bf93cf921ba8225b201a/pkg/reconciler/testing/factory.go#L107-L117) every time and the Defaulting reactor would get called after the `ValidateCreates`.

Reply:
I think the bigger problem with the Reactor route is that the object doesn't get passed along the Reaction chain. So if we have a `SetDefaults` and `ValidateCreates`. Any modifications in SetDefaults won't be reflected when ValidateCreates is called. If SetDefaults returns true for `handled`, it will stop calling any other Reactors which will defeat the purpose of the validation.

Reply:
I'm sending out an update to my other PR that would avoid these changes